rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+/// all writes that were issued at the time that create_checkpoint is called. If force_flush is true,
+/// then SlateDB will force the current wal, or memtable if wal_enabled is false, to flush its data.
+/// Otherwise, the database will wait for the current wal or memtable to be flushed due to
+/// flush_interval or reaching l0_sst_size_bytes, respectively. If set to Durable, then the
+/// checkpoint includes only writes that were durable at the time of the call. This will be faster, but
+/// may not include data from recent writes.
+enum CheckpointScope {
+    All{force_flush: bool},
+    Durable
+}
+
+/// Specify options to provide when creating a checkpoint.
+struct CheckpointOptions {
+    scope: CheckpointScope
+}
+
+impl Db {
+    /// Opens a Db from a checkpoint. If no database This will create a new db under path from
+    /// the checkpoint ID specified in source_checkpoint. SlateDB will look for the manifest
+    /// containing the checkpoint under source_path. The newly created database is a shallow
+    /// copy of the source database, and contains all the data from source_checkpoint. New writes
+    /// will be written to the new database and will not be reflected in the source database.
+    pub async fn open_from_checkpoint(
+        path: Path,
+        object_store: Arc<dyn ObjectStore>,
+        source_path: Path,
+        source_checkpoint: UUID,
+    ) -> Result<Self, SlateDBError> {
+        …
+    }
+
+    /// Create a checkpoint using the provided options. Returns the ID of the created checkpoint.
+    pub async fn create_checkpoint(options: &CheckpointOptions) -> Result<uuid::UUID, SlateDBError> {
+        …
+    }
+
+    /// Called to destroy the database at the given path. If `soft` is true, This method will set the destroyed_at_s field in the manifest. The GC will clean up the db after some time has passed, and all checkpoints have either been deleted or expired. As part of cleaning up the db, the GC will also remove the database’s checkpoint from the source database. If `soft` is false, then all cleanup will be performed by the call to this method. If `soft` is false, the destroy will return SlateDbError::InvalidDeletion if there are any remaining non-expired checkpoints.
+   pub async fn destroy(path: Path, object_store: Arc<dyn ObjectStore>, soft: bool) -> Result<(), SlateDbError> {
+       …
+   }
+}
+
+/// Configuration options for the database reader. These options are set on client startup.
+#[derive(Clone)]
+pub struct DbReaderOptions {
+    /// How frequently to poll for new manifest files. Refreshing the manifest file allows readers to detect newly compacted data.
+    pub manifest_poll_interval: Duration,
+
+
+    /// For readers that refresh their checkpoint, this specifies the lifetime to use for the created checkpoint. The checkpoint’s expire time will be set to the current time plus this value. If not specified, then the checkpoint will be created with no expiry, and must be manually removed.
```

Comment:
Does this also define the frequency for creating checkpoints? I.e. the reader will create a new checkpoint just before expiration of the current one?

Reply:
I think my question is answered below. It sounds like the reader will poll the manifest as specified by `manifest_poll_interval` and establish new checkpoints as needed based on any updates. It might also be useful to define the behavior when the checkpoint expires without any changes detected (say the writer has crashed or is offline for an extended upgrade). Perhaps the reader could just update the expiry time of the current checkpoint when half its lifetime has passed.